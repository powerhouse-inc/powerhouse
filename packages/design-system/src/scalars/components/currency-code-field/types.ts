import type { IconName } from "#powerhouse";
import type React from "react";

export interface Currency {
  ticker: string;
  crypto: boolean;
  label?: string;
  symbol?: string;
  icon?: IconName | React.ComponentType<{ className?: string }>;
}

export type AllowedTypes = "Crypto" | "Fiat" | "Both";
