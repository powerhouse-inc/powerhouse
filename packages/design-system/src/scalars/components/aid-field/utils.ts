import { AutocompleteOption } from "@/scalars/components/fragments/autocomplete-field/types";

export const mockedOptions: AutocompleteOption[] = [
  {
    icon: "Person",
    title: "Agent A",
    path: "agents/agent-a",
    value: "did:ethr:0xb9c5714089478a327f09197987f16f9e5d936e8a",
    description: "Agent A description",
  },
  {
    icon: "Person",
    title: "Agent B",
    path: "agents/agent-b",
    value: "did:ethr:0x5:0xb9c5714089478a327f09197987f16f9e5d936e8a",
    description: "Agent B description",
  },
  {
    icon: "Person",
    title: "Agent C",
    path: "agents/agent-c",
    value: "did:ethr:0x89:0xb9c5714089478a327f09197987f16f9e5d936e8a",
    description: "Agent C description",
  },
];

export const fetchOptions = async (): Promise<AutocompleteOption[]> => {
  // Simulate 2s network delay
  await new Promise((resolve) => setTimeout(resolve, 2000));

  // Add 30% chance of error
  if (Math.random() < 0.3) {
    throw new Error();
  }

  return mockedOptions;
};

export const fetchSelectedOption = async (
  value: string,
): Promise<AutocompleteOption | undefined> => {
  // Simulate 2s network delay
  await new Promise((resolve) => setTimeout(resolve, 2000));
  return mockedOptions.find((option) => option.value === value);
};
