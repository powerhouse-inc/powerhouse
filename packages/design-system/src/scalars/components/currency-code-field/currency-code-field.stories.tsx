import type { Meta, StoryObj } from "@storybook/react";
import { withForm } from "../../lib/decorators.js";
import {
  getDefaultArgTypes,
  getValidationArgTypes,
  StorybookControlCategory,
} from "../../lib/storybook-arg-types.js";
import { CurrencyCodeField } from "./currency-code-field.js";
const meta: Meta<typeof CurrencyCodeField> = {
  title: "Document Engineering/Scalars/Currency Code Field",
  component: CurrencyCodeField,
  decorators: [withForm, (Story) => <div className="w-48">{Story()}</div>],
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  argTypes: {
    ...getDefaultArgTypes(),
    ...getValidationArgTypes(),

    currencies: {
      control: "object",
      description: "Array of Currency objects",
      table: {
        type: { summary: "Currency[]" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    includeCurrencySymbols: {
      control: "boolean",
      description: " Whether to display currency symbols alongside codes",
      table: {
        defaultValue: { summary: "true" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    favoriteCurrencies: {
      control: "object",
      description:
        "List of currencies to display at the top of the dropdown for quick access",
      table: {
        type: { summary: "string[]" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    symbolPosition: {
      control: "select",
      description: "Position of the currency symbol",
      options: ["left", "right"],
      table: {
        type: { summary: "string" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
      if: {
        arg: "includeCurrencySymbols",
        eq: true,
      },
    },
    allowedTypes: {
      control: "select",
      description:
        "Allowed types of currencies to display when no currencies are provided",
      options: ["Fiat", "Crypto", "Both"],
      table: {
        type: { summary: "string" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    searchable: {
      control: "boolean",
      description: "Whether the dropdown is searchable",
      table: {
        defaultValue: { summary: "false" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
  },

  args: {
    name: "currency-code-field",
    placeholder: "Select a currency",
    favoriteCurrencies: [],
    currencies: [],
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    label: "Currency",
    onChange: () => {},
    allowedTypes: "Both",
  },
};

export const Disabled: Story = {
  args: {
    label: "Currency",
    value: "EUR",
    disabled: true,
    onChange: () => {},
    allowedTypes: "Fiat",
  },
};

export const WithFavorites: Story = {
  args: {
    label: "Currency",
    onChange: () => {},
    currencies: [
      {
        ticker: "BTC",
        crypto: true,
        label: "Bitcoin",
        symbol: "₿",
      },
      {
        ticker: "ETH",
        crypto: true,
        label: "Ether",
        symbol: "Ξ",
      },
      {
        ticker: "USDS",
        crypto: true,
        label: "Sky USD",
        symbol: "USDS",
      },
      {
        ticker: "USDC",
        crypto: true,
        icon: "Briefcase",
      },
    ],
    favoriteCurrencies: ["BTC", "ETH"],
  },
};
