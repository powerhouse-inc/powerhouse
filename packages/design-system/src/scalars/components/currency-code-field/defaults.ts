import { type Currency } from "./types.js";

export const commonFiatCurrencies: Currency[] = [
  {
    ticker: "EUR",
    crypto: false,
    label: "Euro",
    symbol: "€",
  },
  {
    ticker: "USD",
    crypto: false,
    label: "United States Dollar",
    symbol: "$",
  },
  {
    ticker: "THB",
    crypto: false,
    label: "Thai Baht",
  },
];

export const commonCryptoCurrencies: Currency[] = [
  {
    ticker: "BTC",
    crypto: true,
    label: "Bitcoin",
    symbol: "₿",
  },
  {
    ticker: "ETH",
    crypto: true,
    label: "Ether",
    symbol: "Ξ",
  },
  {
    ticker: "USDS",
    crypto: true,
    label: "Sky USD",
  },
  {
    ticker: "USDC",
    crypto: true,
  },
];
