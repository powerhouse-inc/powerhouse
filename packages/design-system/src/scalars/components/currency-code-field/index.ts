export { CurrencyCodeField } from "./currency-code-field.js";
export { commonCryptoCurrencies, commonFiatCurrencies } from "./defaults.js";
export type { AllowedTypes, Currency } from "./types.js";
export { cryptoCurrencies, currencies, fiatCurrencies } from "./utils.js";
