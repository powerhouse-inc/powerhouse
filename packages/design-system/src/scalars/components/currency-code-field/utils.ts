// add docs

import { type AllowedTypes } from "./types.js";

/**
 * Get the fiat currencies
 * @returns {Currency[]} The fiat currencies
 */
export const fiatCurrencies = () => [
  { ticker: "USD", crypto: false, label: "USD", symbol: "$" },
  { ticker: "EUR", crypto: false, label: "EUR", symbol: "€" },
  { ticker: "GBP", crypto: false, label: "GBP", symbol: "£" },
];

/**
 * Get the crypto currencies
 * @returns {Currency[]} The crypto currencies
 */
export const cryptoCurrencies = () => [
  { ticker: "DAI", crypto: true, label: "DAI", symbol: "DAI" },
  { ticker: "ETH", crypto: true, label: "ETH", symbol: "ETH" },
  { ticker: "MKR", crypto: true, label: "MKR", symbol: "MKR" },
  { ticker: "SKY", crypto: true, label: "SKY", symbol: "SKY" },
  { ticker: "USDC", crypto: true, label: "USDC", symbol: "USDC" },
  { ticker: "USDS", crypto: true, label: "USDS", symbol: "USDS" },
];

/**
 * Get the currencies
 * @returns {Currency[]} The currencies
 */
export const currencies = () => {
  return [...fiatCurrencies(), ...cryptoCurrencies()];
};

export const getCurrencies = (allowedTypes: AllowedTypes = "Both") => {
  switch (allowedTypes) {
    case "Fiat":
      return fiatCurrencies();
    case "Crypto":
      return cryptoCurrencies();
    case "Both":
      return currencies();
  }
};
