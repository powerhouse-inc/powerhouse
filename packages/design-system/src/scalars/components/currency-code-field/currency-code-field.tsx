import { Select, type SelectOption } from "#ui";
import React, { useMemo } from "react";
import { FormGroup } from "../fragments/index.js";
import { withFieldValidation } from "../fragments/with-field-validation/with-field-validation.js";
import type { FieldErrorHandling, InputBaseProps } from "../types.js";
import type { AllowedTypes, Currency } from "./types.js";
import { getCurrencies } from "./utils.js";

type CurrencyCodeFieldBaseProps = Omit<
  React.ButtonHTMLAttributes<HTMLButtonElement>,
  | keyof InputBaseProps<string | string[]>
  | keyof FieldErrorHandling
  | "onChange"
  | "onBlur"
>;

export interface CurrencyCodeFieldProps
  extends CurrencyCodeFieldBaseProps,
    InputBaseProps<string | string[]>,
    FieldErrorHandling {
  placeholder?: string;
  onChange?: (value: string | string[]) => void;
  onBlur?: () => void;
  currencies?: Currency[];
  includeCurrencySymbols?: boolean;
  favoriteCurrencies?: string[];
  symbolPosition?: "left" | "right";
  searchable?: boolean;
  contentClassName?: string;
  contentAlign?: "start" | "end" | "center";
  allowedTypes?: AllowedTypes;
}

export const CurrencyCodeFieldRaw = React.forwardRef<
  HTMLButtonElement,
  CurrencyCodeFieldProps
>(
  (
    {
      placeholder,
      currencies = [],
      favoriteCurrencies = [],
      includeCurrencySymbols = true,
      symbolPosition = "right",
      searchable = false,
      contentClassName,
      contentAlign = "start",
      allowedTypes = "Both",
      ...props
    },
    ref,
  ) => {
    const defaultCurrencies =
      currencies && currencies.length > 0
        ? currencies
        : getCurrencies(allowedTypes);
    const options: SelectOption[] = useMemo(() => {
      const favoriteTickers = new Set(favoriteCurrencies);

      return (
        (defaultCurrencies
          .map((currency) => {
            if (favoriteTickers.has(currency.ticker)) {
              return null;
            }

            let label = currency.label ?? currency.ticker;
            if (includeCurrencySymbols && currency.symbol) {
              label =
                symbolPosition === "right"
                  ? `${label} (${currency.symbol})`
                  : `(${currency.symbol}) ${label}`;
            }
            const option: SelectOption = {
              label,
              value: currency.ticker,
            };

            if ("icon" in currency) {
              option.icon = currency.icon;
            }

            return option;
          })
          .filter(Boolean) as SelectOption[]) ?? []
      );
    }, [
      defaultCurrencies,
      includeCurrencySymbols,
      symbolPosition,
      favoriteCurrencies,
    ]);

    const favoriteOptions: SelectOption[] = useMemo(() => {
      const favoriteTickers = new Set(favoriteCurrencies);
      return (
        defaultCurrencies
          .filter((currency) => favoriteTickers.has(currency.ticker))
          .map((currency) => {
            let label = currency.label ?? currency.ticker;
            if (includeCurrencySymbols && currency.symbol) {
              label =
                symbolPosition === "right"
                  ? `${label} (${currency.symbol})`
                  : `(${currency.symbol}) ${label}`;
            }
            const option: SelectOption = {
              label,
              value: currency.ticker,
            };

            if ("icon" in currency) {
              option.icon = currency.icon;
            }

            return option;
          }) ?? []
      );
    }, [
      defaultCurrencies,
      favoriteCurrencies,
      includeCurrencySymbols,
      symbolPosition,
    ]);

    return (
      <FormGroup>
        <Select
          ref={ref}
          options={options}
          selectionIcon="checkmark"
          searchable={searchable}
          multiple={false}
          placeholder={placeholder}
          contentAlign={contentAlign}
          contentClassName={contentClassName}
          favoriteOptions={favoriteOptions}
          {...props}
        />
      </FormGroup>
    );
  },
);

CurrencyCodeFieldRaw.displayName = "CurrencyCodeFieldRaw";

export const CurrencyCodeField =
  withFieldValidation<CurrencyCodeFieldProps>(CurrencyCodeFieldRaw);

CurrencyCodeField.displayName = "CurrencyCodeField";
