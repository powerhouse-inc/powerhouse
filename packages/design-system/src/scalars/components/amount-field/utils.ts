import { Currency } from "../currency-code-field";
import { AmountFieldPropsGeneric } from "./types";
export const DEFAULT_FIAT_CURRENCIES: Currency[] = [
  { ticker: "USD", crypto: false, label: "USD", symbol: "$" },
  { ticker: "EUR", crypto: false, label: "EUR", symbol: "€" },
  { ticker: "GBP", crypto: false, label: "GBP", symbol: "£" },
];

export const DEFAULT_CRYPTO_CURRENCIES: Currency[] = [
  { ticker: "DAI", crypto: true, label: "DAI", symbol: "DAI" },
  { ticker: "ETH", crypto: true, label: "ETH", symbol: "ETH" },
  { ticker: "MKR", crypto: true, label: "MKR", symbol: "MKR" },
  { ticker: "SKY", crypto: true, label: "SKY", symbol: "SKY" },
  { ticker: "USDC", crypto: true, label: "USDC", symbol: "USDC" },
  { ticker: "USDS", crypto: true, label: "USDS", symbol: "USDS" },
];

export const DEFAULT_ALL_CURRENCIES: Currency[] = [
  ...DEFAULT_FIAT_CURRENCIES,
  ...DEFAULT_CRYPTO_CURRENCIES,
];
export const getDefaultUnits = (type: AmountFieldPropsGeneric["type"]) => {
  switch (type) {
    case "AmountCurrencyFiat":
      return DEFAULT_FIAT_CURRENCIES;
    case "AmountCurrencyCrypto":
      return DEFAULT_CRYPTO_CURRENCIES;
    case "AmountCurrencyUniversal":
      return DEFAULT_ALL_CURRENCIES;
    default:
      return [];
  }
};

export const isValidBigInt = (value: string | undefined): boolean => {
  if (!value) {
    return false;
  }
  // Ensure the value contains only digits (no decimal points)
  const bigintRegex = /^\d+$/;
  return bigintRegex.test(value);
};

export const isValidNumberGreaterThanMaxSafeInteger = (
  value: string | undefined,
) => {
  if (!value) {
    return false;
  }
  const isValidBigIntValue = isValidBigInt(value);
  return (
    isValidBigIntValue && Math.abs(Number(value)) > Number.MAX_SAFE_INTEGER
  );
};

export const isNotSafeValue = (value: string) => {
  return Math.abs(Number(value)) > Number.MAX_SAFE_INTEGER;
};

export const displayValueAmount = (
  value?: string,
  precision?: number,
  viewPrecision?: number,
  trailingZeros?: boolean,
) => {
  if (!value) {
    return undefined; // Return undefined if no value is provided
  }

  // If viewPrecision is provided but not precision, format to viewPrecision
  if (viewPrecision !== undefined && precision === undefined) {
    const formattedValue = parseFloat(value).toFixed(viewPrecision);
    return trailingZeros
      ? formattedValue
      : parseFloat(formattedValue).toString();
  }

  // If precision is provided but not viewPrecision, truncate to precision
  if (precision !== undefined && viewPrecision === undefined) {
    const formattedValue = parseFloat(value).toFixed(precision);

    // If the number of decimals is greater than or equal to the precision, apply toFixed
    return parseFloat(formattedValue).toString();
  }

  // If both are present, show the value with viewPrecision when focused
  if (precision !== undefined && viewPrecision !== undefined) {
    const formattedValue = parseFloat(value).toFixed(viewPrecision);
    return trailingZeros ? formattedValue : parseFloat(formattedValue);
  }
  return value;
};
