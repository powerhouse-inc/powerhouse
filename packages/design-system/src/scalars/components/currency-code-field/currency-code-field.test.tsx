import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";
import { renderWithForm } from "../../lib/testing.js";
import { Form } from "../form/index.js";
import { CurrencyCodeField } from "./currency-code-field.js";
import { currencies } from "./utils.js";

describe("CurrencyCodeField", () => {
  window.HTMLElement.prototype.scrollIntoView = vi.fn();

  it("should match snapshot", () => {
    const { container } = renderWithForm(
      <CurrencyCodeField
        name="currency"
        label="Currency"
        description="Select a currency"
        value="USD"
        currencies={currencies()}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it("should render with default props", () => {
    renderWithForm(
      <CurrencyCodeField
        name="currency"
        label="Currency"
        currencies={currencies()}
      />,
    );
    expect(screen.getByRole("combobox")).toBeInTheDocument();
    expect(screen.getByText("Currency")).toBeInTheDocument();
  });

  it("should show an error if trying to submit empty with required enabled", async () => {
    const user = userEvent.setup();
    renderWithForm(
      <>
        <CurrencyCodeField
          name="currency"
          label="Currency"
          required
          currencies={currencies()}
        />
        <button type="submit">Submit</button>
      </>,
    );
    await user.click(screen.getByText("Submit"));
    expect(screen.getByText("This field is required")).toBeInTheDocument();
  });

  it("should submit the currency code", async () => {
    const onSubmit = vitest.fn();
    const user = userEvent.setup();
    render(
      <Form onSubmit={onSubmit}>
        <CurrencyCodeField
          name="currency"
          label="Currency"
          value="USD"
          currencies={currencies()}
        />
        <button type="submit">Submit</button>
      </Form>,
    );
    await user.click(screen.getByText("Submit"));
    expect(onSubmit).toHaveBeenCalledWith({ currency: "USD" });
  });

  it("should handle disabled state", () => {
    renderWithForm(
      <CurrencyCodeField
        name="currency"
        label="Currency"
        disabled
        currencies={currencies()}
      />,
    );
    expect(screen.getByRole("combobox")).toBeDisabled();
  });

  it("should display custom errors", async () => {
    const errors = ["Please select a valid currency"];
    renderWithForm(
      <CurrencyCodeField
        name="currency"
        label="Currency"
        errors={errors}
        currencies={currencies()}
      />,
    );
    expect(await screen.findByText(errors[0])).toBeInTheDocument();
  });

  it("should display custom warnings", async () => {
    const warnings = ["This currency may have restrictions"];
    renderWithForm(
      <CurrencyCodeField
        name="currency"
        label="Currency"
        warnings={warnings}
        currencies={currencies()}
      />,
    );
    expect(await screen.findByText(warnings[0])).toBeInTheDocument();
  });
});
