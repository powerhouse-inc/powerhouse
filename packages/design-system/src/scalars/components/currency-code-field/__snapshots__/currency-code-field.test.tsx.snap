// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CurrencyCodeField > should match snapshot 1`] = `
<div>
  <form
    id=":r0:"
    novalidate=""
  >
    <div
      class="flex flex-col gap-2"
    >
      <label
        class="inline-flex items-center text-sm font-medium leading-4 text-gray-900 dark:text-gray-50 mb-[3px]"
        for=":r1:-select"
      >
        Currency
      </label>
      <button
        aria-controls="radix-:r2:"
        aria-expanded="false"
        aria-haspopup="dialog"
        aria-invalid="false"
        class="gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none disabled:pointer-events-none [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 flex h-9 w-full items-center justify-between px-3 py-2 dark:border-charcoal-700 dark:bg-charcoal-900 rounded-md border border-gray-300 bg-white hover:border-gray-300 hover:bg-gray-100 dark:hover:border-charcoal-700 dark:hover:bg-charcoal-800 dark:focus:ring-charcoal-300 focus:outline-none focus:ring-1 focus:ring-gray-900 focus:ring-offset-0 dark:focus-visible:ring-charcoal-300 focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-0"
        data-state="closed"
        id=":r1:-select"
        name="currency"
        role="combobox"
        type="button"
      >
        <div
          class="flex w-full items-center justify-between gap-2"
        >
          <div
            class="max-w-full truncate text-gray-900 dark:text-gray-50 flex items-center gap-2"
          >
            <span
              class="truncate text-[14px] font-normal leading-5"
            />
          </div>
          <div
            class="flex items-center justify-between gap-2"
          >
            <div
              data-testid="icon-fallback"
              style="width: 16px; height: 16px;"
            />
          </div>
        </div>
      </button>
      <p
        class="font-sans text-sm font-normal leading-5 text-gray-600 dark:text-gray-500"
      >
        Select a currency
      </p>
    </div>
  </form>
</div>
`;
